File: src/pyramid.zig
const std = @import("std");

last_inputs: []const f64 = undefined,
fwd_out: []f64,
bkw_out: []f64,
batchSize: usize,
size: usize,

const Self = @This();

pub const threshold: f64 = 1.0;
pub const leak_slope: f64 = 0.01;

pub fn init(
    alloc: std.mem.Allocator,
    batchSize: usize,
    size: usize,
) !Self {
    return Self{
        .last_inputs = try alloc.alloc(f64, size * batchSize),
        .fwd_out = try alloc.alloc(f64, size * batchSize),
        .bkw_out = try alloc.alloc(f64, size * batchSize),
        .batchSize = batchSize,
        .size = size,
    };
}

pub fn forward(self: *Self, inputs: []f64) void {
    std.debug.assert(inputs.len == self.size * self.batchSize);

    var i: usize = 0;
    while (i < inputs.len) : (i += 1) {
        const x = inputs[i];
        if (x < 0) {
            self.fwd_out[i] = leak_slope * x;
        } else if (x < threshold) {
            self.fwd_out[i] = x;
        } else if (x < 2 * threshold) {
            self.fwd_out[i] = 2 * threshold - x;
        } else {
            self.fwd_out[i] = leak_slope * (2 * threshold - x);
        }
    }
    self.last_inputs = inputs;
}

pub fn backwards(self: *Self, grads: []f64) void {
    std.debug.assert(grads.len == self.size * self.batchSize);
    var i: usize = 0;
    while (i < self.last_inputs.len) : (i += 1) {
        const x = self.last_inputs[i];
        if (x < 0) {
            self.bkw_out[i] = leak_slope * grads[i];
        } else if (x < threshold) {
            self.bkw_out[i] = grads[i];
        } else if (x < 2 * threshold) {
            self.bkw_out[i] = -grads[i];
        } else {
            self.bkw_out[i] = -leak_slope * grads[i];
        }
    }
}

File: src/pyramid.java
import java.util.Arrays;

public class Pyramid {
    private double[] lastInputs;
    private double[] fwdOut;
    private double[] bkwOut;
    private int batchSize;
    private int size;

    public static final double THRESHOLD = 1.0;
    public static final double LEAK_SLOPE = 0.01;

    public Pyramid(int batchSize, int size) {
        this.batchSize = batchSize;
        this.size = size;
        this.lastInputs = new double[size * batchSize];
        this.fwdOut = new double[size * batchSize];
        this.bkwOut = new double[size * batchSize];
    }

    public void forward(double[] inputs) {
        if (inputs.length != this.size * this.batchSize) {
            throw new IllegalArgumentException("Input size mismatch");
        }

        for (int i = 0; i < inputs.length; i++) {
            double x = inputs[i];
            if (x < 0) {
                this.fwdOut[i] = LEAK_SLOPE * x;
            } else if (x < THRESHOLD) {
                this.fwdOut[i] = x;
            } else if (x < 2 * THRESHOLD) {
                this.fwdOut[i] = 2 * THRESHOLD - x;
            } else {
                this.fwdOut[i] = LEAK_SLOPE * (2 * THRESHOLD - x);
            }
        }
        this.lastInputs = Arrays.copyOf(inputs, inputs.length);
    }

    public void backwards(double[] grads) {
        if (grads.length != this.size * this.batchSize) {
            throw new IllegalArgumentException("Gradient size mismatch");
        }

        for (int i = 0; i < this.lastInputs.length; i++) {
            double x = this.lastInputs[i];
            if (x < 0) {
                this.bkwOut[i] = LEAK_SLOPE * grads[i];
            } else if (x < THRESHOLD) {
                this.bkwOut[i] = grads[i];
            } else if (x < 2 * THRESHOLD) {
                this.bkwOut[i] = -grads[i];
            } else {
                this.bkwOut[i] = -LEAK_SLOPE * grads[i];
            }
        }
    }

    public double[] getFwdOut() {
        return fwdOut;
    }

    public double[] getBkwOut() {
        return bkwOut;
    }
}
